server/src/main/java/org/opensearch/gateway/AsyncShardFetch.java

```diff
@@ -177,16 +214,26 @@ public synchronized FetchResult<T> fetchData(DiscoveryNodes nodes, Set<String> i
                     }
                 }
             }
-            Set<String> allIgnoreNodes = unmodifiableSet(new HashSet<>(nodesToIgnore));
+
+            Map<ShardId, Set<String>> allIgnoreNodesMap = unmodifiableMap(new HashMap<>(shardToIgnoreNodes));
             // clear the nodes to ignore, we had a successful run in fetching everything we can
             // we need to try them if another full run is needed
-            nodesToIgnore.clear();
+            shardToIgnoreNodes.clear();
             // if at least one node failed, make sure to have a protective reroute
             // here, just case this round won't find anything, and we need to retry fetching data
-            if (failedNodes.isEmpty() == false || allIgnoreNodes.isEmpty() == false) {
-                reroute(shardId, "nodes failed [" + failedNodes.size() + "], ignored [" + allIgnoreNodes.size() + "]");
+            if (failedNodes.isEmpty() == false
+                || allIgnoreNodesMap.values().stream().anyMatch(ignoreNodeSet -> ignoreNodeSet.isEmpty() == false)) {
+                reroute(
+                    logKey,
```

Comment:
Can we rename this to probably `reroutingKey` ?  Also, we can avoid appending readable strings in key itself and consider doing this during logging itself using `ParameterizedMessage`.

Reply:
`reroutingKey` makes sense

Not getting second point, can you please share one example.

Reply:
I am referring to prefixes and suffixes like in this string `"BatchID=[" + batchId + "]"`. We can just log them directly via ParametrizedMessage and just use clean batchId/shardId as the reroutingKey.

Reply:
+1 to above recommendation of using `reroutingKey`. `logKey` is weak logical entity in context of Async Shard Fetch.